actions/query_transform.go

```diff
@@ -0,0 +1,125 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package actions
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+//nolint:lll
+var (
+	dmlVerbs         = []string{"insert", "update", "delete", "replace"}
+	updateRe         = regexp.MustCompile(`(?i)^update\s+(?:low_priority|ignore)?\s*(.*?)\s+set\s+(.*?)(?:\s+where\s+(.*?))?(?:\s+limit\s*[0-9]+(?:\s*,\s*[0-9]+)?)?$`)
+	deleteRe         = regexp.MustCompile(`(?i)^delete\s+(.*?)\bfrom\s+(.*?)$`)
+	insertRe         = regexp.MustCompile(`(?i)^(?:insert(?:\s+ignore)?|replace)\s+.*?\binto\s+(.*?)\(([^\)]+)\)\s*values?\s*\((.*?)\)\s*(?:\slimit\s|on\s+duplicate\s+key.*)?\s*$`)
+	insertReNoFields = regexp.MustCompile(`(?i)^(?:insert(?:\s+ignore)?|replace)\s+.*?\binto\s+(.*?)\s*values?\s*\((.*?)\)\s*(?:\slimit\s|on\s+duplicate\s+key.*)?\s*$`)
+	insertSetRe      = regexp.MustCompile(`(?i)(?:insert(?:\s+ignore)?|replace)\s+(?:.*?\binto)\s+(.*?)\s*set\s+(.*?)\s*(?:\blimit\b|on\s+duplicate\s+key.*)?\s*$`)
+)
+
+func isDMLQuery(query string) bool {
+	query = strings.ToLower(strings.TrimSpace(query))
+	for _, verb := range dmlVerbs {
+		if strings.HasPrefix(query, verb) {
+			return true
+		}
+	}
+	return false
+}
+
+/*
+  MySQL version prior 5.6.3 cannot run explain on DML commands.
+  From the doc: http://dev.mysql.com/doc/refman/5.6/en/explain.html
+  "As of MySQL 5.6.3, permitted explainable statements for EXPLAIN are
+  SELECT, DELETE, INSERT, REPLACE, and UPDATE.
+  Before MySQL 5.6.3, SELECT is the only explainable statement."
+
+  This function converts DML queries to the equivalent SELECT to make
+  it able to explain DML queries on older MySQL versions
+*/
+func dmlToSelect(query string) string {
+	m := updateRe.FindStringSubmatch(query)
+	// > 2 because we need at least a table name and a list of fields
+	if len(m) > 2 {
+		return updateToSelect(m)
+	}
+
+	m = deleteRe.FindStringSubmatch(query)
+	if len(m) > 1 {
```

Comment:
@percona-csalguero Maybe it is better to move these checks into methods itselfs. It will be validated also if you will call this method directly. What do you think? In all cases. It can save a lot of ifs.

Reply:
It is not validating, it is checking if each regular expression matches to choose which function to call.
if `updteRe` matches and returns at least 2 values then call `updateToSelect` and so on.